sql/src/main/java/io/crate/planner/consumer/SemiJoins.java

```diff
@@ -0,0 +1,259 @@
+import io.crate.analyze.relations.RelationNormalizer;
+import io.crate.analyze.symbol.FuncSymbols;
+import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.Literal;
+import io.crate.analyze.symbol.RefReplacer;
+import io.crate.analyze.symbol.SelectSymbol;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.analyze.symbol.SymbolVisitor;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Functions;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.table.Operation;
+import io.crate.operation.operator.OrOperator;
+import io.crate.operation.operator.any.AnyNeqOperator;
+import io.crate.operation.operator.any.AnyOperator;
+import io.crate.operation.predicate.NotPredicate;
+import io.crate.planner.node.dql.join.JoinType;
+import io.crate.sql.tree.QualifiedName;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+import static io.crate.analyze.expressions.ExpressionAnalyzer.castIfNeededOrFail;
+import static io.crate.operation.operator.Operators.LOGICAL_OPERATORS;
+import static io.crate.operation.scalar.cast.CastFunctionResolver.isCastFunction;
+
+final class SemiJoins {
+
+    private final RelationNormalizer relationNormalizer;
+
+    public SemiJoins(Functions functions) {
+        relationNormalizer = new RelationNormalizer(functions);
+    }
+
+    /**
+     * Try to rewrite a QueriedRelation into a SemiJoin:
+     *
+     * <pre>
+     *
+     *     select x from t1 where x in (select id from t2)
+     *               |
+     *               v
+     *     select t1.x from t1 SEMI JOIN (select id from t2) t2 on t1.x = t2.id
+     * </pre>
+     *
+     * (Note that it's not possible to write a SemiJoin directly using SQL. But the example above is semantically close.)
+     *
+     * This rewrite isn't always possible - certain conditions need to be met for the semi-join to have the same
+     * semantics as the subquery.
+     *
+     * @return the rewritten relation or null if a rewrite wasn't possible.
+     */
+    @Nullable
+    QueriedRelation tryRewrite(QueriedRelation rel, TransactionContext transactionCtx) {
+        WhereClause where = rel.querySpec().where();
+        if (!where.hasQuery()) {
+            return null;
+        }
+        List<Function> rewriteCandidates = gatherRewriteCandidates(where.query());
+        if (rewriteCandidates.isEmpty()) {
+            return null;
+        }
+        if (rewriteCandidates.size() > 1) {
+            return null;  // TODO: support this as well
+        }
+        Function rewriteCandidate = rewriteCandidates.get(0);
+        SelectSymbol selectSymbol = getSubqueryOrNull(rewriteCandidate.arguments().get(1));
+        assert selectSymbol != null : "rewriteCandidate must contain a selectSymbol";
```

Comment:
Is this assert still valid? The above `getSubqueryOrNull` seems to indicate `null` is a reasonable value